File: src/ui/editordrawer/automationprojection.cpp
#include "ui/editordrawer/automationprojection.h"

#include <algorithm>
#include <cmath>

#include "core/timedefaults.h"
#include "ui/editordrawer/automationpage.h"
#include "ui/layout.h"

AutomationGeometry AutomationGeometry::resolve()
{
    AutomationGeometry geometry;
    geometry.rowDefaultHeight = layout::fontPx(4.0);
    geometry.rowMinimumHeight = layout::fontPx(7.0 / 3.0);
    geometry.rowMaximumHeight = layout::fontPx(32.0 / 3.0);
    geometry.rowWheelIncrement = layout::fontPx(1.0 / 3.0);
    geometry.addLaneStripHeight = layout::fontPx(5.0 / 3.0);
    geometry.plotOrigin = layout::fontPx(17.5 + 13.0 / 3.0);
    geometry.pointHitRadius = layout::fontPx(7.0 / 12.0);
    geometry.neutralSnapRadius = layout::fontPx(2.0 / 3.0);
    geometry.deleteTimeRadius = layout::fontPx(3.0 / 4.0);
    geometry.pointDetailThreshold = layout::fontPx(2.0);
    geometry.hoverPaintPadding = layout::fontPx(1.0 / 6.0);
    geometry.nodeDragActivationDistance = layout::fontPx(5.0 / 12.0);
    geometry.gridMinimumCellWidth = layout::fontPx(4.0 / 3.0);
    geometry.nodePaintRadius = layout::fontPxF(3.0 / 16.0);
    geometry.selectedNodeRingRadius = layout::fontPxF(9.0 / 32.0);
    geometry.nodeOutlineDipWidth = layout::fontPxF(1.0 / 12.0);
    geometry.selectedNodeRingDipWidth = layout::fontPxF(1.0 / 5.0);
    // Plot endpoint centers at the painted outer edge, including marker stroke.
    const qreal nodeOuterRadius = geometry.nodePaintRadius + geometry.nodeOutlineDipWidth;
    const qreal selectedNodeOuterRadius =
        geometry.selectedNodeRingRadius + geometry.selectedNodeRingDipWidth * 0.5;
    geometry.valuePlotPadding = qRound(std::max(nodeOuterRadius, selectedNodeOuterRadius));
    return geometry;
}

int AutomationProjection::rowHeight(const AutomationRow &row) const
{
    if (!m_page)
        return m_geometry.rowDefaultHeight;
    return std::clamp(m_page->laneHeightFor(row.id), m_geometry.rowMinimumHeight,
                      m_geometry.rowMaximumHeight);
}

int AutomationProjection::rowTop(int index) const
{
    int top = m_topInset;
    for (int row = 0; row < index && row < int(m_rows.size()); ++row)
        top += rowHeight(m_rows[row]);
    return top;
}

std::pair<int, int> AutomationProjection::valuePlotBounds(int index) const
{
    const int top = rowTop(index);
    const int height = rowHeight(m_rows[index]);
    return {top + m_geometry.valuePlotPadding, top + height - m_geometry.valuePlotPadding};
}

int AutomationProjection::rowIndexAt(int y) const
{
    int bottom = m_topInset;
    for (int row = 0; row < int(m_rows.size()); ++row) {
        bottom += rowHeight(m_rows[row]);
        if (y < bottom)
            return row;
    }
    return -1;
}

int AutomationProjection::rowBoundaryAt(int y) const
{
    int bottom = m_topInset;
    for (int row = 0; row < int(m_rows.size()); ++row) {
        bottom += rowHeight(m_rows[row]);
        if (std::abs(y - bottom) <= layout::singlePixel())
            return row;
    }
    return -1;
}

int AutomationProjection::rowMinimum(const AutomationRow &row) const
{
    if (row.id.kind != EditorAutomationRowKind::ControlChange)
        return CoreTimeDefaults::kMinCcValue;
    return CoreTimeDefaults::laneValueMinimum(row.id.controller);
}

int AutomationProjection::rowMaximum(const AutomationRow &row) const
{
    if (row.id.kind != EditorAutomationRowKind::ControlChange)
        return CoreTimeDefaults::kMaxCcValue;
    if (row.id.controller == automation::kBendController ||
        !automation::rangeZoomable(row.id.controller))
        return CoreTimeDefaults::laneValueMaximum(row.id.controller);
    int dataMaximum = 0;
    if (const auto *lane = laneFor(row))
        for (const auto &point : lane->points)
            dataMaximum = std::max(dataMaximum, point.value);
    const auto it = m_page->m_viewState.laneRanges.find(row.id);
    const uint8_t mode = it == m_page->m_viewState.laneRanges.cend()
                             ? automation::defaultRange(row.id.controller)
                             : it->second;
    return mode == 0 ? automation::autoRange(dataMaximum) : std::max(int(mode), dataMaximum);
}

int AutomationProjection::valueAtY(int rowIndex, qreal y) const
{
    if (rowIndex < 0 || rowIndex >= int(m_rows.size()))
        return 0;
    const auto [top, bottom] = valuePlotBounds(rowIndex);
    const int minimum = rowMinimum(m_rows[rowIndex]);
    const int maximum = rowMaximum(m_rows[rowIndex]);
    const qreal clamped = std::clamp(y, qreal(top), qreal(bottom));
    return minimum + int((qreal(bottom) - clamped) * qreal(maximum - minimum) /
                         qreal(std::max(1, bottom - top)));
}

qreal AutomationProjection::pointY(const AutomationRow &row, int rowIndex, int value) const
{
    const auto [top, bottom] = valuePlotBounds(rowIndex);
    return qreal(bottom) - qreal(value - rowMinimum(row)) * qreal(bottom - top) /
                               qreal(std::max(1, rowMaximum(row) - rowMinimum(row)));
}

double AutomationProjection::rawTickAt(qreal x) const
{
    return std::max(0.0, m_page->tickAtContentX(std::max(qreal(m_geometry.plotOrigin), x) -
                                                m_geometry.plotOrigin));
}
qreal AutomationProjection::displayX(uint64_t tick, qreal devicePixelRatio) const
{
    return m_page->displayX(tick, m_geometry.plotOrigin, devicePixelRatio);
}

uint64_t AutomationProjection::snapTickAt(qreal x, bool fine) const
{
    return m_page->snapTick(rawTickAt(x), fine);
}

qreal AutomationProjection::valueY(const QRect &bounds, const AutomationGeometry &geometry,
                                   double minimum, double maximum, double value)
{
    const qreal top = bounds.top() + geometry.valuePlotPadding;
    const qreal bottom = bounds.bottom() - geometry.valuePlotPadding;
    const double clamped = std::clamp(value, minimum, maximum);
    return bottom -
           qreal((clamped - minimum) * (bottom - top) / std::max<qreal>(1.0, maximum - minimum));
}

double AutomationProjection::valueAtY(const QRect &bounds, const AutomationGeometry &geometry,
                                      double minimum, double maximum, qreal y)
{
    const qreal top = bounds.top() + geometry.valuePlotPadding;
    const qreal bottom = bounds.bottom() - geometry.valuePlotPadding;
    const qreal clamped = std::clamp(y, top, bottom);
    return minimum +
           double(bottom - clamped) * (maximum - minimum) / std::max<qreal>(1.0, bottom - top);
}

uint64_t AutomationProjection::fineSnapTick(double rawTick) const
{
    return m_page->snapTick(rawTick, true);
}

const AutoLane *AutomationProjection::laneFor(const AutomationRow &row) const
{
    if (row.id.kind != EditorAutomationRowKind::ControlChange)
        return nullptr;
    return m_page->model().findLane(int(row.id.track), row.id.controller);
}

bool AutomationProjection::nodeMarkersVisible() const
{
    return m_page && m_page->pxPerBeat() >= m_geometry.pointDetailThreshold;
}

AutomationProjection::PointerMapping AutomationProjection::pointerMapping(int rowIndex, qreal x,
                                                                          qreal y) const
{
    PointerMapping mapped;
    if (!m_page || !m_page->timeline() || rowIndex < 0 || rowIndex >= int(m_rows.size()))
        return mapped;
    const uint64_t length = m_page->timeline()->lengthTicks;
    mapped.rawTick = std::clamp(rawTickAt(x), 0.0, double(length));
    mapped.point.value = valueAtY(rowIndex, y);
    mapped.cell = snapCellAt(mapped.rawTick);
    mapped.point.tick = mapped.cell.tickBegin;
    return mapped;
}

AutomationGridCell AutomationProjection::snapCellAt(double rawTick) const
{
    if (!m_page || !m_page->timeline() || m_page->timeline()->lengthTicks == 0)
        return {};
    const uint64_t length = m_page->timeline()->lengthTicks;
    const double clamped = std::clamp(rawTick, 0.0, double(length));
    const uint64_t tick = clamped >= double(length) ? length - 1 : uint64_t(std::floor(clamped));
    const uint64_t start = m_page->snapTickDown(double(tick), false);
    return {start, m_page->nextGridTick(start, false, length)};
}

const std::vector<AutomationGridCell> &
AutomationProjection::snapCellsCrossed(std::vector<AutomationGridCell> &cells,
                                       double previousRawTick, double currentRawTick) const
{
    cells.clear();
    if (!m_page || !m_page->timeline() || m_page->timeline()->lengthTicks == 0)
        return cells;
    const uint64_t length = m_page->timeline()->lengthTicks;
    const bool forward = currentRawTick >= previousRawTick;
    const auto target = snapCellAt(currentRawTick);
    auto cell = snapCellAt(previousRawTick);
    while (cell.tickBegin < cell.tickEnd && cell.tickEnd <= length) {
        cells.push_back(cell);
        if (cell.tickBegin == target.tickBegin)
            return cells;
        if (forward) {
            const uint64_t nextStart = m_page->nextGridTick(cell.tickBegin, false, length);
            if (nextStart >= length)
                break;
            const auto next = snapCellAt(double(nextStart));
            if (next.tickBegin <= cell.tickBegin)
                break;
            cell = next;
        } else {
            if (cell.tickBegin == 0)
                break;
            const auto previous = snapCellAt(double(cell.tickBegin - 1));
            if (previous.tickBegin >= cell.tickBegin)
                break;
            cell = previous;
        }
    }
    cells.clear();
    return cells;
}

File: src/swift/app/drawer/automation/AutomationProjection.swift
import Foundation
import PorydawCore

// The pure Automation projection domain: the font-relative interaction
// geometry, the grid/clock snapping lattice, and the per-event projection
// engine over the shared camera. Every x is plot-relative (0 = the plot's
// left edge); the page publishes the plot origin separately.

// MARK: - Geometry and snapping

/// One half-open visible grid cell. A tick on a boundary belongs to the cell
/// that begins there, which is the single boundary rule for snapping, pencil
/// cells and hit testing.
public struct AutomationGridCell: Equatable, Sendable {
    public let tickBegin: Tick
    public let tickEnd: Tick

    public init(tickBegin: Tick, tickEnd: Tick) {
        self.tickBegin = tickBegin
        self.tickEnd = tickEnd
    }

    public func contains(_ rawTick: Double) -> Bool {
        rawTick >= Double(tickBegin) && rawTick < Double(tickEnd)
    }
}

/// `AutomationGeometry::resolve`: the font-relative interaction geometry.
public struct AutomationPlotGeometry: Equatable, Sendable {
    public let rowHeight: Double
    public let pointHitRadius: Double
    public let neutralSnapRadius: Double
    public let nodeDragActivationDistance: Double
    public let pointDetailThreshold: Double
    public let valuePlotPadding: Double

    public init(baseFontPx: Double) {
        let base = baseFontPx.isFinite && baseFontPx > 0
            ? baseFontPx
            : GridCameraPolicy.seedBaseFontPx
        let nodePaintRadius = fontPxF(base, 3.0 / 16.0)
        let nodeOutlineDipWidth = fontPxF(base, 1.0 / 12.0)
        let selectedRingRadius = fontPxF(base, 9.0 / 32.0)
        let selectedRingDipWidth = fontPxF(base, 1.0 / 5.0)
        rowHeight = fontPx(base, 4.0)
        pointHitRadius = fontPx(base, 7.0 / 12.0)
        neutralSnapRadius = fontPx(base, 2.0 / 3.0)
        nodeDragActivationDistance = fontPx(base, 5.0 / 12.0)
        pointDetailThreshold = fontPx(base, 2.0)
        // Endpoint centers sit at the painted outer edge, marker stroke included.
        valuePlotPadding = (max(nodePaintRadius + nodeOutlineDipWidth,
                                selectedRingRadius + selectedRingDipWidth * 0.5)).rounded()
    }
}

/// The plot rectangle the automation body owns, in the page's own units.
public struct AutomationPlotBounds: Equatable, Sendable {
    public var width: Double
    public var height: Double
    public var devicePixelRatio: Double

    public init(width: Double = 0, height: Double = 0, devicePixelRatio: Double = 1) {
        self.width = max(0, width.isFinite ? width : 0)
        self.height = max(0, height.isFinite ? height : 0)
        self.devicePixelRatio = devicePixelRatio.isFinite && devicePixelRatio > 0
            ? devicePixelRatio : 1
    }
}

/// The snapping lattice: the shared visible grid for a coarse pointer position,
/// and the shared document clock lattice for a fine (Alt) one.
public struct AutomationSnapPolicy: Sendable {
    private let metrics: GridMetrics
    public let clockTicks: Tick

    init(baseFontPx: Double, devicePixelRatio: Double, timeAxis: TimeAxis,
         clockTicks: Tick) {
        metrics = GridMetrics(baseFontPx: baseFontPx, dpr: devicePixelRatio, width: 0, height: 0,
                              timeAxis: timeAxis)
        self.clockTicks = max(1, clockTicks)
    }

    @MainActor
    public init(document: SongDocument, timeline: PlaybackTimeline, baseFontPx: Double,
                devicePixelRatio: Double) {
        self.init(baseFontPx: baseFontPx, devicePixelRatio: devicePixelRatio,
                  timeAxis: TimeAxis(map: TimeMap(
                      ticksPerBeat: UInt32(max(1, document.ticksPerBeat)),
                      lengthTicks: timeline.lengthTicks,
                      loopStartTick: timeline.loopStartTick,
                      loopEndTick: timeline.loopEndTick,
                      timeSigs: document.timeSignatures.map {
                          TimeSigPoint(tick: $0.tick, numerator: $0.numerator,
                                       denomPow2: $0.denominatorPower)
                      })),
                  clockTicks: TimelineSnapPolicy.clockTicks(
                      division: document.ticksPerBeat,
                      extendedClocks: document.state.config.extendedClocks))
    }


    public func snap(_ tick: Double, fine: Bool, camera: EditorCamera) -> Tick {
        fine ? TimelineSnapPolicy.fineSnap(tick, clockTicks: clockTicks)
             : Tick(metrics.snapTick(tick, camera: camera))
    }

    public func snapDown(_ tick: Double, fine: Bool, camera: EditorCamera) -> Tick {
        let position = max(0, tick)
        guard fine else { return Tick(metrics.snapTickDown(position, camera: camera)) }
        let limit = Double(TimeDefaults.maxTick)
        let clamped = min(position, limit)
        return Tick(min(limit, (clamped / Double(clockTicks)).rounded(.down) * Double(clockTicks)))
    }

    /// `Grid::nextSnapTickAfter`, clamped to `limit`.
    public func next(after tick: Tick, fine: Bool, limit: Tick, camera: EditorCamera) -> Tick {
        if tick >= limit { return limit }
        if fine {
            let next = (UInt64(tick) / UInt64(clockTicks) + 1) * UInt64(clockTicks)
            return min(limit, Tick(min(next, UInt64(TimeDefaults.maxTick))))
        }
        let candidate = metrics.snapTickUp(Double(tick) + 1, camera: camera)
        return min(limit, Tick(max(Int(tick) + 1, candidate)))
    }
}

// MARK: - Projection

/// Per-event projection over the shared camera: tick to x, value to y, the
/// snapping lattice, and cell traversal. Construct one per pointer event or
/// build pass, exactly as production does.
public struct AutomationProjection: Sendable {
    public let camera: EditorCamera
    public let bounds: AutomationPlotBounds
    public let geometry: AutomationPlotGeometry
    public let snapPolicy: AutomationSnapPolicy
    public let songEndTick: Tick
    public let displayMaximum: Int?

    public static func displayMaximum(snapshot: AutomationLaneSnapshot, range: Int?) -> Int? {
        guard snapshot.metadata.zoomable else { return nil }
        let range = range ?? Int(AutomationCatalog.defaultRange(snapshot.parameter.controller ?? 0))
        guard range == 0 else { return range }
        let maximum = snapshot.displaySeries.map(\.value).max() ?? 0
        return Int(AutomationCatalog.autoRange(maximum: maximum))
    }

    public init(camera: EditorCamera, bounds: AutomationPlotBounds,
                geometry: AutomationPlotGeometry, snapPolicy: AutomationSnapPolicy,
                songEndTick: Tick, displayMaximum: Int? = nil) {
        self.camera = camera
        self.bounds = bounds
        self.geometry = geometry
        self.snapPolicy = snapPolicy
        self.songEndTick = songEndTick
        self.displayMaximum = displayMaximum
    }

    // MARK: Time

    public func contentX(_ tick: Tick) -> Double { camera.contentX(tick: Double(tick)) }

    /// The display x of a tick: the camera's projection, snapped to the physical
    /// pixel grid exactly as the roll and every other page snap theirs.
    public func x(_ tick: Tick) -> Double {
        let value = camera.contentX(tick: Double(tick))
        let scale = bounds.devicePixelRatio
        return scale > 0 ? (value * scale).rounded() / scale : value
    }

    /// The raw (unsnapped) tick under an x, clamped to `[0, songEnd]`.
    public func rawTick(atX x: Double) -> Double {
        min(max(0, camera.tickAtContentX(max(0, x))), Double(songEndTick))
    }

    public func tick(atX x: Double, fine: Bool) -> Tick {
        snapPolicy.snap(rawTick(atX: x), fine: fine, camera: camera)
    }

    public func nextGridTick(after tick: Tick, fine: Bool) -> Tick {
        snapPolicy.next(after: tick, fine: fine, limit: songEndTick, camera: camera)
    }

    /// The half-open cell a raw tick lands in. The song's end tick belongs to the
    /// cell that ends there, so a position at the end never starts a new one.
    public func cell(atRawTick rawTick: Double) -> AutomationGridCell {
        guard songEndTick > 0 else { return AutomationGridCell(tickBegin: 0, tickEnd: 0) }
        let clamped = min(max(0, rawTick), Double(songEndTick))
        let position = clamped >= Double(songEndTick) ? songEndTick - 1 : Tick(clamped)
        let begin = snapPolicy.snapDown(Double(position), fine: false, camera: camera)
        return AutomationGridCell(tickBegin: begin,
                                  tickEnd: nextGridTick(after: begin, fine: false))
    }

    /// Every cell the pointer crossed from `previous` to `current`, in traversal
    /// order and including both endpoint cells. Stops at the same guards as
    /// production: a cell that cannot advance, or a walk past the song's end.
    public func cellsCrossed(from previous: Double, to current: Double) -> [AutomationGridCell] {
        guard songEndTick > 0 else { return [] }
        let forward = current >= previous
        let target = cell(atRawTick: current)
        var crossed: [AutomationGridCell] = []
        var current = cell(atRawTick: previous)
        while current.tickBegin < current.tickEnd && current.tickEnd <= songEndTick {
            crossed.append(current)
            if current.tickBegin == target.tickBegin { return crossed }
            if forward {
                let nextStart = nextGridTick(after: current.tickBegin, fine: false)
                if nextStart >= songEndTick { break }
                let next = cell(atRawTick: Double(nextStart))
                if next.tickBegin <= current.tickBegin { break }
                current = next
            } else {
                if current.tickBegin == 0 { break }
                let previousCell = cell(atRawTick: Double(current.tickBegin - 1))
                if previousCell.tickBegin >= current.tickBegin { break }
                current = previousCell
            }
        }
        return crossed
    }

    // MARK: Value

    /// The value axis: the parameter maximum sits at the top edge of the padded
    /// plot, the minimum at the bottom.
    public func y(_ value: Int, metadata: AutomationParameterMetadata) -> Double {
        let top = geometry.valuePlotPadding
        let bottom = max(top, bounds.height - geometry.valuePlotPadding)
        let clamped = Double(metadata.clamp(value))
        let span = max(1, (displayMaximum ?? metadata.maximum) - metadata.minimum)
        return bottom - (clamped - Double(metadata.minimum)) * (bottom - top) / Double(span)
    }

    public func value(atY y: Double, metadata: AutomationParameterMetadata) -> Int {
        let top = geometry.valuePlotPadding
        let bottom = max(top, bounds.height - geometry.valuePlotPadding)
        let clamped = min(max(y, top), bottom)
        let span = Double((displayMaximum ?? metadata.maximum) - metadata.minimum)
        let exact = Double(metadata.minimum)
            + (bottom - clamped) * span / max(1.0, bottom - top)
        return metadata.clamp(Int(exact.rounded()))
    }

    /// `AutomationProjection::nodeMarkersVisible`: node markers only at a zoom
    /// that can show them.
    public func markersVisible() -> Bool {
        camera.snapshot.pixelsPerBeat >= geometry.pointDetailThreshold
    }

    // MARK: Lane projection

    /// Project one frozen lane snapshot. Pure: nothing here mutates the document.
    public func project(_ snapshot: AutomationLaneSnapshot,
                        selection: AutomationTimeSelection? = nil,
                        usedTracks: Set<Int> = []) -> AutomationLaneProjection {
        let metadata = snapshot.metadata
        let covers = selection?.covers(snapshot.parameter, usedTracks: usedTracks) ?? false
        let range = covers ? selection?.range : nil

        func projected(_ tick: Tick, _ value: Int, identity: AutomationPointIdentity,
                       projectedNode: Bool) -> AutomationProjectedPoint {
            AutomationProjectedPoint(
                identity: identity, tick: tick, value: value,
                x: x(tick), y: y(value, metadata: metadata),
                selected: range.map { $0.contains(tick) } ?? false,
                projected: projectedNode)
        }

        // Display points: the synthetic tick-zero node first, then the written
        // occurrences with the last at a tick winning, as the lane adapters do.
        let points: [AutomationProjectedPoint] = snapshot.displaySeries.map { item in
            projected(item.tick, item.value, identity: item.identity,
                      projectedNode: item.projected)
        }

        let leadIn = snapshot.leadInValue.map {
            projected(0, metadata.clamp($0),
                      identity: AutomationPointIdentity(
                          revision: snapshot.revision, parameter: snapshot.parameter, tick: 0,
                          occurrence: -1, value: metadata.clamp($0)),
                      projectedNode: true)
        }

        // The step/ramp curve: each point holds (or ramps) into the next, the
        // last one runs to the song's end, and the lead-in opens the lane.
        var segments: [AutomationCurveSegment] = []
        if let leadIn, let first = points.first, leadIn.value != first.value {
            segments.append(AutomationCurveSegment(
                kind: .step, tickBegin: 0, tickEnd: first.tick, fromValue: leadIn.value,
                toValue: leadIn.value, isLeadIn: true, isSelected: false))
        }
        for (index, point) in points.enumerated() {
            let next = index + 1 < points.count ? points[index + 1] : nil
            let kind: AutomationCurveSegment.Kind =
                metadata.interpolation == .ramp ? .ramp : .step
            segments.append(AutomationCurveSegment(
                kind: kind, tickBegin: point.tick, tickEnd: next?.tick,
                fromValue: point.value,
                toValue: kind == .ramp ? (next?.value ?? point.value) : point.value,
                isLeadIn: false, isSelected: point.selected))
        }

        return AutomationLaneProjection(
            parameter: snapshot.parameter, metadata: metadata, revision: snapshot.revision,
            points: points, sources: snapshot.sources, leadIn: leadIn, segments: segments,
            scaleLabels: scaleLabels(metadata: metadata), songEndTick: snapshot.songEndTick,
            selectionRange: range)
    }

    /// Live modifier mapping against the captured time/value projection.
    func mappedPoint(x: Double, y: Double, facts: AutomationFrozenFacts,
                     modifiers: AutomationModifiers, plotHeight: Double,
                     neutralSnapRadius: Double) -> AutomationLanePoint {
        AutomationLanePoint(
            tick: tick(atX: x, fine: modifiers.fine),
            value: facts.metadata.snappedValue(
                value(atY: y, metadata: facts.metadata), snapValue: modifiers.snapValue,
                plotHeight: plotHeight, neutralSnapRadius: neutralSnapRadius))
    }

    /// The highlighted scale labels: maximum, minimum, and — only when the
    /// parameter has a neutral value — the neutral one, all at curve-true Y.
    public func scaleLabels(metadata: AutomationParameterMetadata) -> [AutomationScaleLabel] {
        let maximum = displayMaximum ?? metadata.maximum
        var labels: [AutomationScaleLabel] = [
            AutomationScaleLabel(role: .maximum, value: maximum,
                                 text: metadata.valueText(maximum),
                                 y: y(maximum, metadata: metadata)),
            AutomationScaleLabel(role: .minimum, value: metadata.minimum,
                                 text: metadata.valueText(metadata.minimum),
                                 y: y(metadata.minimum, metadata: metadata)),
        ]
        if let neutral = metadata.neutral {
            labels.append(AutomationScaleLabel(role: .neutral, value: neutral,
                                               text: metadata.valueText(neutral),
                                               y: y(neutral, metadata: metadata)))
        }
        return labels
    }
}
